File: AddTwoNumbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
# Input: (2 -> 4 -> 3) + (5 -> 6 -> 4)
# Output: 7 -> 0 -> 8
# Explanation: 342 + 465 = 807.
class Solution:
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        dummyhead = ListNode(0)
        temp1 = l1
        temp2 = l2
        current = dummyhead
        sum = 0
        while(temp1 != None or temp2 != None):
            if temp1 != None:
                sum += temp1.val
                temp1 = temp1.next
            if temp2 != None:
                sum += temp2.val
                temp2 = temp2.next

            current.next = ListNode(sum%10)
            sum //= 10
            current = current.next

        if sum == 1:
            current.next = ListNode(1)

        return dummyhead.next

File: AddTwoNumbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        int carry;
        int sums;
        int x,y;
        int result;
        carry = 0;
        ListNode head = new ListNode(0);
        ListNode current = head;
        while(l1 != null || l2 != null){
            if(l1 != null){
                x = l1.val;
            }else{
                x = 0;
            }
            if(l2 != null){
                y = l2.val;
            }else{
                y = 0;
            }
            sums = x + y + carry;
            result = sums%10;
            current.next = new ListNode(result);
            current = current.next;
            carry = sums/10;
            if(l1 != null){
                l1 = l1.next;
            }
            if(l2 != null){
                l2 = l2.next;
            }
        }
        if(carry > 0){
            ListNode temp = new ListNode(1);
            current.next = temp;
        }
        return head.next;

    }
}
